### ISSUE TITLE
Questions about pbench-fio

### ISSUE THREAD

**Issue by hongkailiu** [distributed-system-analysis/pbench#654]
I am new to pbench-fio and met the following problems:

Command 1 finises in about 2 minutes:
```sh
# pbench-fio --test-types=read --clients=172.20.1.6 --config=SEQ_IO --samples=1 --max-stddev=20 --block-sizes=4 --job-file=config/sequential_io.job 
```

Command 2
If --test-types=rw, then it seems hanging there after I see "fio job complete" line in the output.

The client is a pod based on this image: https://github.com/openshift/svt/blob/master/storage/fio/Dockerfile

Questions:
About command 1: The file in the log is not generated: /var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T19.07.53/1-read-4KiB/sample1/clients/172.20.1.6/fio_clat_hist.1.log
Is this about some setting I did not make right?
About command 2: What caused long-time running?
The output has the job file for fio. Is it possible to get the fio command pbench uses?

Thanks.
Hongkai

-------

```sh
# yum list installed | grep pbench
configtools.noarch              0.3.1-1           @copr-pbench                  
pbench-agent.noarch             0.44-1gf694c2f    @copr-pbench                  
pbench-sysstat.x86_64           11.2.0-1          @copr-pbench

# pbench-fio --test-types=read --clients=172.20.1.6 --config=SEQ_IO --samples=1 --max-stddev=20 --block-sizes=4 --job-file=config/sequential_io.job 
verifying clients have fio installed
[error][2017-08-30T19:07:53.614861319] Unable to find default tools group directory, /var/lib/pbench-agent/tools-default
Collecting system information
The following jobfile was created: /var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T19.07.53/1-read-4KiB/sample1/fio.job
[global]
direct=0
sync=0
fsync_on_close=1
time_based=1
runtime=30
clocksource=clock_gettime
ramp_time=5
startdelay=5
filename=/var/lib/fio/test
size=1g
write_bw_log=fio
write_iops_log=fio
write_lat_log=fio
write_hist_log=fio
per_job_logs=1
log_avg_msec=60000
log_hist_msec=60000

[fio-1]
bs=4k
rw=read
numjobs=1

running fio job: /var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T19.07.53/1-read-4KiB/sample1/fio.job
bash: firewall-cmd: command not found
Unable to determine bin values from input clat_hist files. Namely 
the first line of file '/var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T19.07.53/1-read-4KiB/sample1/clients/172.20.1.6/fio_clat_hist.1.log' has 1 
columns of which we assume -2 correspond to histogram bins. 
This number needs to be equal to one of the following numbers:

     ([[1216, 1280, 1344, 1408, 1472, 1536, 1600, 1664],
       [ 608,  640,  672,  704,  736,  768,  800,  832],
       [ 304,  320,  336,  352,  368,  384,  400,  416],
       [ 152,  160,  168,  176,  184,  192,  200,  208],
       [  76,   80,   84,   88,   92,   96,  100,  104],
       [  38,   40,   42,   44,   46,   48,   50,   52],
       [  19,   20,   21,   22,   23,   24,   25,   26],
       [ N/A,   10,  N/A,   11,  N/A,   12,  N/A,   13],
       [ N/A,    5,  N/A,  N/A,  N/A,    6,  N/A,  N/A]])

Possible reasons and corresponding solutions:
  - Input file(s) does not contain histograms.
  - You recompiled fio with a different GROUP_NR. If so please specify this
    new GROUP_NR on the command line with --group_nr
Chart Type: xy
fio job complete
Use of uninitialized value $metric_section in split at /opt/pbench-agent/bench-scripts/postprocess/generate-benchmark-summary line 290.
Use of uninitialized value $metric_section in printf at /opt/pbench-agent/bench-scripts/postprocess/generate-benchmark-summary line 294.
Use of uninitialized value $text in split at /opt/pbench-agent/bench-scripts/postprocess/BenchPostprocess.pm line 76.
Use of uninitialized value $metric_section in printf at /opt/pbench-agent/bench-scripts/postprocess/generate-benchmark-summary line 340.
Use of uninitialized value $metric_section in split at /opt/pbench-agent/bench-scripts/postprocess/generate-benchmark-summary line 477.
Use of uninitialized value $metric_section in printf at /opt/pbench-agent/bench-scripts/postprocess/generate-benchmark-summary line 481.
Collecting system information



# pbench-fio --test-types=rw --clients=172.20.1.6 --config=SEQ_IO --samples=1 --max-stddev=20 --block-sizes=4 --job-file=config/sequential_io.job 
verifying clients have fio installed
[error][2017-08-30T19:11:33.119879863] Unable to find default tools group directory, /var/lib/pbench-agent/tools-default
Collecting system information
The following jobfile was created: /var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T19.11.32/1-rw-4KiB/sample1/fio.job
[global]
direct=0
sync=0
fsync_on_close=1
time_based=1
runtime=30
clocksource=clock_gettime
ramp_time=5
startdelay=5
filename=/var/lib/fio/test
size=1g
write_bw_log=fio
write_iops_log=fio
write_lat_log=fio
write_hist_log=fio
per_job_logs=1
log_avg_msec=60000
log_hist_msec=60000

[fio-1]
bs=4k
rw=rw
numjobs=1

running fio job: /var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T19.11.32/1-rw-4KiB/sample1/fio.job
bash: firewall-cmd: command not found
Unable to determine bin values from input clat_hist files. Namely 
the first line of file '/var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T19.11.32/1-rw-4KiB/sample1/clients/172.20.1.6/fio_clat_hist.1.log' has 1 
columns of which we assume -2 correspond to histogram bins. 
This number needs to be equal to one of the following numbers:

     ([[1216, 1280, 1344, 1408, 1472, 1536, 1600, 1664],
       [ 608,  640,  672,  704,  736,  768,  800,  832],
       [ 304,  320,  336,  352,  368,  384,  400,  416],
       [ 152,  160,  168,  176,  184,  192,  200,  208],
       [  76,   80,   84,   88,   92,   96,  100,  104],
       [  38,   40,   42,   44,   46,   48,   50,   52],
       [  19,   20,   21,   22,   23,   24,   25,   26],
       [ N/A,   10,  N/A,   11,  N/A,   12,  N/A,   13],
       [ N/A,    5,  N/A,  N/A,  N/A,    6,  N/A,  N/A]])

Possible reasons and corresponding solutions:
  - Input file(s) does not contain histograms.
  - You recompiled fio with a different GROUP_NR. If so please specify this
    new GROUP_NR on the command line with --group_nr
Chart Type: xy
fio job complete
```

**Comment by hongkailiu** [comment]
@atheurer @bengland2

**Comment by atheurer** [comment]
Looks like the histogram postprocessing is failing?  @bengland2 is there a way to disable this processing?  @hongkailiu can you remove "log_hist_msec=60000" from your job file and try again?

**Comment by bengland2** [comment]
So don't run [fiologparser_hist.py](https://github.com/axboe/fio/blob/master/tools/hist/fiologparser_hist.py).   pbench-fio is just a shell script, you can comment out postprocessing [right here](https://github.com/distributed-system-analysis/pbench/blob/master/agent/bench-scripts/pbench-fio#L607), I'd guess.  Looks like the input histogram log may be malformed.

**Comment by hongkailiu** [comment]
I tried both. Neither worked.
Tried also adding debug output in the shell script.
It struck with this line:
$benchmark_results_dir/$benchmark-postprocess.cmd

The actual command in this file is:
```sh
# cat /var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T23.26.02/1-rw-4KiB/sample1/fio-postprocess.cmd
/opt/pbench-agent/bench-scripts/postprocess/fio-postprocess "/var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T23.26.02/1-rw-4KiB/sample1" "fio-" "default"
```

For some reason, its rc is 130 and did not generate the result.json file. And then iteration kicked off.

```
Can't open /var/lib/pbench-agent/fio_SEQ_IO_2017.08.30T23.26.02/1-rw-4KiB/sample1/result.json: No such file or directory at /opt/pbench-agent/bench-scripts/postprocess/process-iteration-samples line 174.
Restarting iteration 1-rw-4KiB (1 of 1)
```

How should I debug further?

Thanks.
Hongkai

**Comment by atheurer** [comment]
Can you upload that result to the pbench web repository so we can look at the output of that run?

**Comment by hongkailiu** [comment]
I am not sure where the pbench repo is and sent you an email with the pbench results attached.

**Comment by hongkailiu** [comment]
@atheurer 
Any progress on this issue? Thanks.

**Comment by hongkailiu** [comment]
More info:
The following change makes it work.
log_avg_msec=60000
to
log_avg_msec=10000

I do not understand the cause but hopefully info gives a clue of something. :)

**Comment by bengland2** [comment]
oh yeah, that's the sample rate for the fio logs, you reduced it from 60
sec to 10 sec, makes sense.   It's averaging IOPS, latency, etc. over that
interval.   If you don't specify this parameter, then fio will try to
output every single I/O request's latency, which is huge and you don't want
that (use histogram logs instead).  There is also a "log_hist_msec"
parameter for histogram logs.  Tim Wilkinson and I had to run some tests
that took like 9 hours to do, so we increased the interval between samples
to minimize the amount of data that we'd have to collect at the end (from
hundreds of VMs).  This might be relevant to container work also.   But
your tests are much shorter so you don't have to worry about this as much,
and you need to get some samples to get statistics.

**Comment by bengland2** [comment]
I added a pbench-fio --histogram_interval_msec option so you don't have to use the fio job file to generate histogram logs for short tests.    @ndokos Can we close this one?

**Comment by portante** [comment]
Long over due to be closed given @bengland2's original work.

**Comment by portante** [comment]
Fixed by #746 and #958.
